statistics/builder.go

```diff
@@ -263,51 +261,56 @@ func (t *Table) buildColumn(sc *variable.StatementContext, offset int, ndv int64
 	bucketIdx := 0
 	var lastNumber int64
 	for i := int64(0); i < int64(len(samples)); i++ {
-		cmp, err := col.Values[bucketIdx].CompareDatum(sc, samples[i])
+		cmp, err := col.Buckets[bucketIdx].Value.CompareDatum(sc, samples[i])
 		if err != nil {
 			return errors.Trace(err)
 		}
 		if cmp == 0 {
 			// The new item has the same value as current bucket value, to ensure that
 			// a same value only stored in a single bucket, we do not increase bucketIdx even if it exceeds
 			// valuesPerBucket.
-			col.Numbers[bucketIdx] = i * sampleFactor
-			col.Repeats[bucketIdx] += sampleFactor
+			col.Buckets[bucketIdx].Count = (i + 1) * sampleFactor
+			col.Buckets[bucketIdx].Repeats += sampleFactor
 		} else if i*sampleFactor-lastNumber <= valuesPerBucket {
+			// TODO: Making sampleFactor as float may be better.
 			// The bucket still have room to store a new item, update the bucket.
-			col.Numbers[bucketIdx] = i * sampleFactor
-			col.Values[bucketIdx] = samples[i]
-			col.Repeats[bucketIdx] = 0
+			col.Buckets[bucketIdx].Count = (i + 1) * sampleFactor
+			col.Buckets[bucketIdx].Value = samples[i]
+			col.Buckets[bucketIdx].Repeats = sampleFactor
 		} else {
 			// The bucket is full, store the item in the next bucket.
-			lastNumber = col.Numbers[bucketIdx]
```

Comment:
I think we need a test to make sure that the number of buckets do not exceed the predefined bucket count.